// @nrwl/storybook/src/utils/testing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMockContext = exports.getTestArchitect = exports.createTestUILib = exports.runMigration = exports.callRule = exports.runSchematic = void 0;
const tslib_1 = require("tslib");
const path_1 = require("path");
const os_1 = require("os");
const fs_1 = require("fs");
const core_1 = require("@angular-devkit/core");
const schematics_1 = require("@angular-devkit/schematics");
const testing_1 = require("@angular-devkit/schematics/testing");
const architect_1 = require("@angular-devkit/architect");
const testing_2 = require("@angular-devkit/architect/testing");
const testing_3 = require("@nrwl/workspace/testing");
const testRunner = new testing_1.SchematicTestRunner('@nrwl/storybook', path_1.join(__dirname, '../../collection.json'));
testRunner.registerCollection('@nrwl/angular', path_1.join(__dirname, '../../../angular/collection.json'));
testRunner.registerCollection('@nrwl/react', path_1.join(__dirname, '../../../react/collection.json'));
testRunner.registerCollection('@nrwl/jest', path_1.join(__dirname, '../../../jest/collection.json'));
testRunner.registerCollection('@nrwl/cypress', path_1.join(__dirname, '../../../cypress/collection.json'));
const migrationRunner = new testing_1.SchematicTestRunner('@nrwl/storybook/migrations', path_1.join(__dirname, '../../migrations.json'));
function runSchematic(schematicName, options, tree) {
    return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}
exports.runSchematic = runSchematic;
function callRule(rule, tree) {
    return testRunner.callRule(rule, tree).toPromise();
}
exports.callRule = callRule;
function runMigration(migrationName, options, tree) {
    return migrationRunner
        .runSchematicAsync(migrationName, options, tree)
        .toPromise();
}
exports.runMigration = runMigration;
function createTestUILib(libName, collectionName, options = {}) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        let appTree = schematics_1.Tree.empty();
        appTree = testing_3.createEmptyWorkspace(appTree);
        appTree = yield callRule(schematics_1.externalSchematic(collectionName, 'library', Object.assign({ name: libName }, options)), appTree);
        appTree = yield callRule(schematics_1.externalSchematic(collectionName, 'component', {
            name: 'test-button',
            project: libName,
        }), appTree);
        if (collectionName === '@nrwl/angular') {
            updateAngularComponent(appTree);
        }
        if (collectionName === '@nrwl/react') {
            // @TODO
        }
        appTree = yield callRule(schematics_1.externalSchematic(collectionName, 'component', {
            name: 'test-other',
            project: libName,
        }), appTree);
        return appTree;
        function updateAngularComponent(appTree) {
            appTree.overwrite(`libs/${libName}/src/lib/test-button/test-button.component.ts`, `
import { Component, OnInit, Input } from '@angular/core';
import { tmpdir } from 'os';
import { mkdtempSync } from 'fs';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent implements OnInit {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age: number;
  @Input() isOn = false;

  constructor() { }

  ngOnInit() {
  }

}
`);
            appTree.overwrite(`libs/${libName}/src/lib/test-button/test-button.component.html`, `<button [attr.type]="type" [ngClass]="style"></button>`);
        }
    });
}
exports.createTestUILib = createTestUILib;
function getTempDir() {
    const tmpDir = os_1.tmpdir();
    const tmpFolder = `${tmpDir}${path_1.sep}`;
    return fs_1.mkdtempSync(tmpFolder);
}
function getTestArchitect() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const tmpDir = getTempDir();
        const architectHost = new testing_2.TestingArchitectHost(tmpDir, tmpDir);
        const registry = new core_1.schema.CoreSchemaRegistry();
        registry.addPostTransform(core_1.schema.transforms.addUndefinedDefaults);
        const architect = new architect_1.Architect(architectHost, registry);
        yield architectHost.addBuilderFromPackage(path_1.join(__dirname, '../..'));
        return [architect, architectHost];
    });
}
exports.getTestArchitect = getTestArchitect;
function getMockContext() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const [architect, architectHost] = yield getTestArchitect();
        const context = new testing_3.MockBuilderContext(architect, architectHost);
        yield context.addBuilderFromPackage(path_1.join(__dirname, '../..'));
        return context;
    });
}
exports.getMockContext = getMockContext;
