// @nrwl/devkit/src/generators/project-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readProjectConfiguration = exports.updateProjectConfiguration = exports.addProjectConfiguration = void 0;
const workspace_1 = require("@nrwl/tao/src/shared/workspace");
const read_json_1 = require("../utils/read-json");
const get_workspace_layout_1 = require("../utils/get-workspace-layout");
/**
 * Adds project configuration to the Nx workspace.
 *
 * The project configuration is stored in workspace.json and nx.json. The utility will update
 * both files.
 *
 * @param host - the file system tree
 * @param projectName - unique name. Often directories are part of the name (e.g., mydir-mylib)
 * @param projectConfiguration - project configuration
 */
function addProjectConfiguration(host, projectName, projectConfiguration) {
    setProjectConfiguration(host, projectName, projectConfiguration, 'create');
}
exports.addProjectConfiguration = addProjectConfiguration;
/**
 * Updates the configuration of an existing project.
 *
 * The project configuration is stored in workspace.json and nx.json. The utility will update
 * both files.
 *
 * @param host - the file system tree
 * @param projectName - unique name. Often directories are part of the name (e.g., mydir-mylib)
 * @param projectConfiguration - project configuration
 */
function updateProjectConfiguration(host, projectName, projectConfiguration) {
    setProjectConfiguration(host, projectName, projectConfiguration, 'update');
}
exports.updateProjectConfiguration = updateProjectConfiguration;
/**
 * Reads a project configuration.
 *
 * The project configuration is stored in workspace.json and nx.json. The utility will read
 * both files.
 *
 * @param host - the file system tree
 * @param projectName - unique name. Often directories are part of the name (e.g., mydir-mylib)
 */
function readProjectConfiguration(host, projectName) {
    return Object.assign(Object.assign({}, readWorkspaceSection(host, projectName)), { nxJsonSection: readNxJsonSection(host, projectName) });
}
exports.readProjectConfiguration = readProjectConfiguration;
function readWorkspaceSection(host, projectName) {
    const path = get_workspace_layout_1.getWorkspacePath(host);
    const workspaceJson = read_json_1.readJson(host, path);
    const newFormat = workspace_1.toNewFormat(workspaceJson);
    if (!newFormat.projects[projectName]) {
        throw new Error(`Cannot find Project '${projectName}'`);
    }
    return newFormat.projects[projectName];
}
function readNxJsonSection(host, projectName) {
    const nxJson = read_json_1.readJson(host, 'nx.json');
    return nxJson.projects[projectName];
}
function setProjectConfiguration(host, projectName, projectConfiguration, mode) {
    addProjectToWorkspaceJson(host, projectName, Object.assign(Object.assign({}, projectConfiguration), { tags: undefined, implicitDependencies: undefined }), mode);
    addProjectToNxJson(host, projectName, {
        tags: projectConfiguration.tags,
        implicitDependencies: projectConfiguration.implicitDependencies,
    });
}
function addProjectToWorkspaceJson(host, projectName, project, mode) {
    const path = get_workspace_layout_1.getWorkspacePath(host);
    const workspaceJson = read_json_1.readJson(host, path);
    if (mode == 'create' && workspaceJson.projects[projectName]) {
        throw new Error(`Cannot create Project '${projectName}'. It already exists.`);
    }
    if (mode == 'update' && !workspaceJson.projects[projectName]) {
        throw new Error(`Cannot update Project '${projectName}'. It does not exist.`);
    }
    workspaceJson.projects[projectName] = project;
    host.write(path, JSON.stringify(workspaceJson));
}
function addProjectToNxJson(host, projectName, config) {
    const nxJson = read_json_1.readJson(host, 'nx.json');
    nxJson.projects[projectName] = Object.assign({
        tags: [],
    }, (config || {}));
    host.write('nx.json', JSON.stringify(nxJson));
}
